### ISSUE TITLE
Package programs have relative import fails when compiled

### ISSUE THREAD

**Issue by jonorthwash** [Nuitka/Nuitka#1330]
When I run create an executable with `nuitka --standalone`, the resulting executable gives errors about relative imports.

* Nuitka version:
```shell
> python -m nuitka --version
0.6.18.4
Commercial: None
Python: 3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)]
Flavor: Unknown
Executable: C:\Users\Jonathan\AppData\Local\Programs\Python\Python39\python.exe
OS: Windows
Arch: x86_64
```
* Installed from `Nuitka-6.1.184.win-amd64.py39.msi`

* Code that I tried to compile: https://github.com/jonorthwash/randomtest

* Output of `nuitka --standalone MyApp`:
```shell
>nuitka --standalone MyApp
Nuitka-Options:INFO: Used command line options: --standalone MyApp
Nuitka:INFO: Starting Python compilation with Nuitka '0.6.18.4' on Python '3.9' commercial None.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'tkinter' added '_tkinter'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'socket' added '_socket'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'ctypes' added '_ctypes'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl).
Nuitka-Scons:INFO: Compiled 10 C files using clcache with 0 cache hits and 10 cache misses.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\Grammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\PatternGrammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka:INFO: Keeping build directory 'MyApp.build'.
Nuitka:INFO: Successfully created 'MyApp.dist\\MyApp.exe'.
```
* Output of running the script via `python -m MyApp` (**expected**):
```shell
>python -m MyApp
testing modules
running module
running util.tools.Tool from within module
init Tool
testing util
init Tool
```
* Output of running executable (**unexpected**):
```shell
>MyApp.dist\MyApp.exe
Traceback (most recent call last):
  File "Z:\2021-12-20\MyApp.dist\__main__.py", line 3, in <module>
ImportError: attempted relative import with no known parent package
```

This issue may be related to #493 or #869.  I'm probably just doing something wrong, but @kayhayen suggested in `Nuitka-chat/community` that I submit an issue about this.

**Comment by kayhayen** [comment]
Thanks, it seems a comment in the closed #869 spells it out that for `__main__` there are different modes, maybe. I need to investigate what `__package__` should be for directory execution. Seems that needs to make a difference, and potentially not all modes of Nuitka do it correctly.

**Comment by kayhayen** [comment]
Ok, seems while normal `__main__` has `__package__` as None, this one has it as empty string `""` which is bizarre, this must be special treated too.

**Comment by kayhayen** [comment]
Actually, there is code in Nuitka that does this, 
```
        "is_dunder_main": 1
        if module_name == "__main__"
        and os.path.basename(module.getCompileTimeFilename()) == "__main__.py"
        else 0,
````

That in turn enables this code:

```
#if %(is_dunder_main)s
        UPDATE_STRING_DICT0(
            moduledict_%(module_identifier)s,
            (Nuitka_StringObject *)const_str_plain___package__,
            const_str_empty
        );
```

But maybe the compile time filename has changed here.

**Comment by kayhayen** [comment]
I managed to reproduce the issue with your repo, simply using accelerated mode with both Python2 and Python3.

**Comment by kayhayen** [comment]
And I did so with CPython as well, so that explains why Nuitka is behaving correctly. Were you to compile it as a package, it would run fine, but that compiles the `__init__` and would be bad. I now sort of understand why the issue is not easy, or at least the solution to put the empty string is not correct.

Problem is also, that being a namespace package of Python3, there is no way of guessing what mode is correct. I need to think about this. An extra option controlling it could be done, but nobody is going to find it, and making it required sort of breaks stuff.

**Comment by kayhayen** [comment]
Ok, I have found the interface as least, it will be `--python-flag=-m` that controls this. Nuitka will still ask for the directory as a filename, but when given `-m`, that will change the value of `__package__` accordingly.

**Comment by kayhayen** [comment]
I pushed to factory a change that makes this work. The only horror I have is how to explain this to people.

**Comment by jonorthwash** [comment]
> I pushed to factory a change that makes this work. 

I'm having trouble with this.  I installed factory, and here's what I get:

Compilation:
```
>nuitka --standalone MyApp
Nuitka-Options:INFO: Used command line options: --standalone MyApp
Nuitka:INFO: Starting Python compilation with Nuitka '0.6.19rc5' on Python '3.9' commercial None.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'tkinter' added '_tkinter'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'socket' added '_socket'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'ctypes' added '_ctypes'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl).
Nuitka-Scons:INFO: Compiled 10 C files using clcache with 0 cache hits and 10 cache misses.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\Grammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\PatternGrammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka:INFO: Keeping build directory 'MyApp.build'.
Nuitka:INFO: Successfully created 'MyApp.dist\\MyApp.exe'.
```

Running:
```
>MyApp.dist\MyApp.exe
Traceback (most recent call last):
  File "C:\Users\Jonathan\2021-1~1\MYAPP~1.DIS\__main__.py", line 3, in <module>
ModuleNotFoundError: No module named 'modules'
```

**Comment by jonorthwash** [comment]
A different result, when I use `--include-plugin-directory`:

Compilation:
```shell
>nuitka --standalone --include-plugin-directory=MyApp\modules MyApp
Nuitka-Options:INFO: Used command line options: --standalone --include-plugin-directory=MyApp\modules MyApp
Nuitka:INFO: Starting Python compilation with Nuitka '0.6.19rc5' on Python '3.9' commercial None.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'tkinter' added '_tkinter'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'socket' added '_socket'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'ctypes' added '_ctypes'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl).
Nuitka-Scons:INFO: Compiled 12 C files using clcache with 12 cache hits and 0 cache misses.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\Grammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\PatternGrammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka:INFO: Keeping build directory 'MyApp.build'.
Nuitka:INFO: Successfully created 'MyApp.dist\\MyApp.exe'.
```

Running:
```shell
>MyApp.dist\MyApp.exe
Traceback (most recent call last):
  File "C:\Users\Jonathan\2021-1~1\MYAPP~1.DIS\__main__.py", line 3, in <module>
  File "C:\Users\Jonathan\2021-1~1\MYAPP~1.DIS\modules\__init__.py", line 1, in <module modules>
  File "C:\Users\Jonathan\2021-1~1\MYAPP~1.DIS\modules\base.py", line 1, in <module modules.base>
ImportError: attempted relative import beyond top-level package
```
```

**Comment by jonorthwash** [comment]
> it will be `--python-flag=-m` that controls this.

I tried this with both approaches above, and get the same results as I reported there.

**Comment by kayhayen** [comment]
@jonorthwash in the one you put, you didn't use it apparently, and the plugin directory option is just wrong to use.

**Comment by kayhayen** [comment]
Follow https://nuitka.net/doc/factory.html for installation, I have observed that people fail to install things properly and updated the instructions recently so that it will.

**Comment by jonorthwash** [comment]
> Follow https://nuitka.net/doc/factory.html for installation, I have observed that people fail to install things properly and updated the instructions recently so that it will.

Yes, I installed it using `git clone --branch factory https://github.com/Nuitka/Nuitka.git`.  Does `0.6.19rc5` not reflect that?  It's up from `0.6.18.4`, which I had before.

**Comment by jonorthwash** [comment]
> @jonorthwash in the one you put, you didn't use it apparently, and the plugin directory option is just wrong to use.

Okay, so I reported this wrong: the output of running is _a little_ different.
```shell
Z:\2021-12-20>nuitka --standalone --python-flag=-m MyApp
Nuitka-Options:INFO: Used command line options: --standalone --python-flag=-m MyApp
Nuitka:INFO: Starting Python compilation with Nuitka '0.6.19rc5' on Python '3.9' commercial None.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'tkinter' added '_tkinter'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'socket' added '_socket'.
Nuitka-Plugins:INFO: implicit-imports: Implicit dependencies of module 'ctypes' added '_ctypes'.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl).
Nuitka-Scons:INFO: Compiled 10 C files using clcache with 0 cache hits and 10 cache misses.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\Grammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka-Plugins:INFO: data-files: Included data file 'lib2to3\PatternGrammar3.9.9.final.0.pickle' due to package data for 'lib2to3.pgen2'.
Nuitka:INFO: Keeping build directory 'MyApp.build'.
Nuitka:INFO: Successfully created 'MyApp.dist\\MyApp.exe'.

Z:\2021-12-20>MyApp.dist\MyApp.exe
Traceback (most recent call last):
  File "Z:\2021-12-20\MyApp.dist\__main__.py", line 3, in <module>
ModuleNotFoundError: No module named 'MyApp'
```

**Comment by kayhayen** [comment]
Ah, that is a shortcoming of the current implementation, indeed, it doesn't follow the imports. So I smuggled it into the MyApp package, but the module object still has full name `__main__` rather than `MyApp.__name__`, so when it compile time analyzes imports they will not be resolved and stand incorrect.

The main module having that full name `__main__` is a deeply rooted assumption in Nuitka. However, I will have to visit it I guess. I made a hack to not include its `__package.__main__` when encoutering it. I probably should make the real solution there.

Also I was thinking if it discoveres `main_added`, if there should not rather be something that makes people disable it. It seems very likely that people want it, and the wrong `__package__` value would only be relevant in case of Python2 and relative or absolute imports, both being tried, but not with Python3, where absolute imports are the default.

For now, you could use `--include-package=MyApp` which effectively disables compile time detection. I think I use that and had it work, but maybe only because I didn't use standalone mode.

**Comment by kayhayen** [comment]
I have started the cleanup, e.g. `MainProgram.c` is now getting the `"__main__"` string passed through a define. There is going to be a lot of these, mostly because I also haven't always checked for `isMainModule()` where it would have been possible. This is going to take some time to finish though.

**Comment by jonorthwash** [comment]
> For now, you could use `--include-package=MyApp` which effectively disables compile time detection.

Yes, this works.  Thank you.

**Comment by kayhayen** [comment]
This will take a bit longer, I have some things for it, but it's not complete until all things are done.

**Comment by kayhayen** [comment]
I have added a test and adapted the test frame work to support `-m` and during that I found that dunder main detection wasn't yet perfect, but probably good enough, as it didn't change much when I made it better.

What I discovered is that in Python3, the module repr of `package.__main__` is only `__main__` despite it now having a `__package__` and `__name__` that are being compatible. And with Python2, it complained it cannot import relatively, because the package is not loaded. And that is one and the same issue. The main program C code imports now the dunder main, but that doesn't import the package it is contained it. That is of course very true. I added a print to the package, and indeed, for package programs, the package gets executed of course first, and only then the main submodule is loaded.

In Nuitka I think there were assumptions that the main module is running first, but I might have removed these before for multiprocessing and other uses, so that should not be an issue. I guess, all it takes is an extra step to load not one, but two modules, check error exit, etc. And to do that as well for the other modes. This might be easy enough, so I can give it a stab right now even before new release.

**Comment by kayhayen** [comment]
So, on factory there is something now that passes my test, and I think that means it ought to be good. I think that somehow `sys.modules[__name__]` is something rather strange and might even be wrong, but that's not what people use anyway a lot, would need to investigate what's going on there. Pretty sure, Nuitka is setting up the module correctly, but like I said, there is dirty stuff going on with how the module is named towards itself and publicly.

https://nuitka.net/doc/factory.html

**Comment by kayhayen** [comment]
Part of the 0.6.19 release I just made.

**Comment by kayhayen** [comment]
I included it in 0.6.19 still, taking the risk of breakage.
